core/worker.go

```diff
@@ -575,29 +589,54 @@ func (w *worker) GeneratePendingHeader(block *types.WorkObject, fill bool) (*typ
-			blockReward := new(big.Int).Add(coinbaseReward, blockFees)
-			coinbaseEtx := types.NewTx(&types.ExternalTx{To: &coinbase, Value: blockReward, IsCoinbase: true, OriginatingTxHash: origin, ETXIndex: 0, Sender: coinbase, Data: []byte{work.wo.Lock()}})
-			work.etxs[0] = coinbaseEtx
-		} else if coinbase.IsInQuaiLedgerScope() {
+			blockReward := new(big.Int).Add(coinbaseReward, work.quaiFees)
+			coinbaseEtx := types.NewTx(&types.ExternalTx{To: &primaryCoinbase, Gas: params.TxGas, Value: blockReward, EtxType: types.CoinbaseType, OriginatingTxHash: origin, ETXIndex: uint16(len(work.etxs)), Sender: primaryCoinbase, Data: []byte{defaultCoinbaseLockup}})
+			work.etxs = append(work.etxs, coinbaseEtx)
+			if work.utxoFees.Cmp(big.NewInt(0)) != 0 {
+				coinbaseEtx := types.NewTx(&types.ExternalTx{To: &secondaryCoinbase, Gas: params.TxGas, Value: work.utxoFees, EtxType: types.CoinbaseType, OriginatingTxHash: origin, ETXIndex: uint16(len(work.etxs)), Sender: secondaryCoinbase, Data: []byte{defaultCoinbaseLockup}})
+				work.etxs = append(work.etxs, coinbaseEtx)
+			}
+		} else if bytes.Equal(work.wo.PrimaryCoinbase().Bytes(), qiCoinbase.Bytes()) {
 			coinbaseReward := misc.CalculateReward(work.wo.WorkObjectHeader())
-			blockFees := new(big.Int).Add(work.quaiFees, misc.QiToQuai(primeTerminus.WorkObjectHeader(), work.utxoFees))
-			blockReward := new(big.Int).Add(coinbaseReward, blockFees)
-			coinbaseEtx := types.NewTx(&types.ExternalTx{To: &coinbase, Value: blockReward, IsCoinbase: true, OriginatingTxHash: origin, ETXIndex: 0, Sender: coinbase, Data: []byte{work.wo.Lock()}})
-			work.etxs[0] = coinbaseEtx
+			blockReward := new(big.Int).Add(coinbaseReward, work.utxoFees)
+			coinbaseEtx := types.NewTx(&types.ExternalTx{To: &primaryCoinbase, Gas: params.TxGas, Value: blockReward, EtxType: types.CoinbaseType, OriginatingTxHash: origin, ETXIndex: uint16(len(work.etxs)), Sender: primaryCoinbase, Data: []byte{defaultCoinbaseLockup}})
+			work.etxs = append(work.etxs, coinbaseEtx)
+			if work.quaiFees.Cmp(big.NewInt(0)) != 0 {
+				coinbaseEtx := types.NewTx(&types.ExternalTx{To: &secondaryCoinbase, Gas: params.TxGas, Value: work.quaiFees, EtxType: types.CoinbaseType, OriginatingTxHash: origin, ETXIndex: uint16(len(work.etxs)), Sender: secondaryCoinbase, Data: []byte{defaultCoinbaseLockup}})
+				work.etxs = append(work.etxs, coinbaseEtx)
+			}
+		}
+
+		// Add an etx for each workshare for it to be rewarded
+		for _, uncle := range uncles {
+			reward := misc.CalculateReward(uncle)
+			uncleCoinbase := uncle.PrimaryCoinbase()
+			work.etxs = append(work.etxs, types.NewTx(&types.ExternalTx{To: &uncleCoinbase, Gas: params.TxGas, Value: reward, EtxType: types.CoinbaseType, OriginatingTxHash: origin, ETXIndex: uint16(len(work.etxs)), Sender: uncleCoinbase, Data: []byte{uncle.Lock()}}))
 		}
+
+	}
+
+	if !fromOrderedTransactionSet {
+		go w.OrderTransactionSet(work.txs, work.gasUsedAfterTransaction, block)
```

Comment:
Why is this done async? We need to be very careful here as this could lead to a goroutine leak

Reply:
Also need to add the defer panic callback

Reply:
Update: depending on how often you call Generate this will cause a leak because it's O(n^2) plus another generate

Reply:
will make it a channel  and that way we can limit how many of these processes can run at a time

Reply:
This is done as well